File: src/services/multipath.ts
/**
 * 多路径路由服务
 */

import { invoke } from '@tauri-apps/api/core'

export type SlicingStrategy =
  | 'RoundRobin'
  | 'Random'
  | 'Weighted'
  | 'LeastConnections'
  | 'LatencyBased'

export type PoolType =
  | 'General'
  | 'Streaming'
  | 'Gaming'
  | 'Download'
  | 'Social'

export interface MultipathConfig {
  enabled: boolean
  strategy: SlicingStrategy
  node_pools: NodePool[]
  min_fragment_size: number
  max_fragment_size: number
  reassembly_timeout: number
  session_persistence: boolean
  bindings: SessionBinding[]
}

export interface NodePool {
  name: string
  pool_type: PoolType
  nodes: PathNode[]
  enabled: boolean
}

export interface PathNode {
  name: string
  server: string
  port: number
  protocol: string
  weight: number
  enabled: boolean
  location?: string
  max_connections?: number
}

export interface SessionBinding {
  domain_pattern: string
  pool_type: PoolType
  force_single_node: boolean
  description: string
}

export interface TestResult {
  success: boolean
  latency: number
  message: string
}

export interface ImportResult {
  success: boolean
  imported_count: number
  message: string
}

export async function multipathGetConfig(): Promise<MultipathConfig> {
  return invoke<MultipathConfig>('multipath_get_config')
}

export async function multipathUpdateConfig(
  config: MultipathConfig,
): Promise<void> {
  return invoke<void>('multipath_update_config', { config })
}

export async function multipathGetBindings(): Promise<SessionBinding[]> {
  return invoke<SessionBinding[]>('multipath_get_bindings')
}

export async function multipathAddBinding(
  binding: SessionBinding,
): Promise<void> {
  return invoke<void>('multipath_add_binding', { binding })
}

export async function multipathRemoveBinding(
  domainPattern: string,
): Promise<void> {
  return invoke<void>('multipath_remove_binding', { domainPattern })
}

export async function multipathGetPredefinedBindings(): Promise<
  SessionBinding[]
> {
  return invoke<SessionBinding[]>('multipath_get_predefined_bindings')
}

export async function multipathAddPool(pool: NodePool): Promise<void> {
  return invoke<void>('multipath_add_pool', { pool })
}

export async function multipathRemovePool(poolName: string): Promise<void> {
  return invoke<void>('multipath_remove_pool', { poolName })
}

export async function multipathUpdatePool(pool: NodePool): Promise<void> {
  return invoke<void>('multipath_update_pool', { pool })
}

export async function multipathAddNode(
  poolName: string,
  node: PathNode,
): Promise<void> {
  return invoke<void>('multipath_add_node', { poolName, node })
}

export async function multipathRemoveNode(
  poolName: string,
  nodeName: string,
): Promise<void> {
  return invoke<void>('multipath_remove_node', { poolName, nodeName })
}

export async function multipathTestNode(node: PathNode): Promise<TestResult> {
  return invoke<TestResult>('multipath_test_node', { node })
}

export async function multipathImportNodes(
  poolName: string,
  nodesYaml: string,
): Promise<ImportResult> {
  return invoke<ImportResult>('multipath_import_nodes', {
    poolName,
    nodesYaml,
  })
}

export async function multipathExportNodes(poolName: string): Promise<string> {
  return invoke<string>('multipath_export_nodes', { poolName })
}

export async function multipathGetRecommendedConfig(): Promise<MultipathConfig> {
  return invoke<MultipathConfig>('multipath_get_recommended_config')
}

File: src-tauri/src/feat/multipath.rs
use crate::multipath::{
    MultipathConfig, MultipathManager, NodePool, PathNode, PoolType, SessionBinding, SlicingStrategy, NodeStats,
};
use anyhow::Result;
use std::collections::HashMap;
use std::sync::Arc;

fn multipath_manager() -> Arc<MultipathManager> {
    crate::feat::get_coordinator().multipath_manager()
}

fn persist_multipath_config(config: &MultipathConfig) -> Result<()> {
    let path = crate::config::AdvancedConfig::default_path()?;
    let mut advanced_config = crate::config::AdvancedConfig::load(&path)?;
    advanced_config.multipath = config.clone();
    advanced_config.save(&path)?;

    let coordinator = crate::feat::get_coordinator();
    coordinator.apply_advanced_config(&advanced_config)?;
    Ok(())
}

pub fn apply_multipath_config(config: MultipathConfig) -> Result<()> {
    let manager = multipath_manager();
    manager.update_config(config.clone());
    persist_multipath_config(&config)?;
    log::info!("[Multipath] config updated");
    Ok(())
}

/// 获取多路径配置
pub fn multipath_get_config() -> MultipathConfig {
    let manager = multipath_manager();
    manager.get_config()
}

/// 获取会话绑定规则
pub fn multipath_get_bindings() -> Vec<SessionBinding> {
    let manager = multipath_manager();
    manager.get_bindings()
}

/// 添加会话绑定规则
pub fn multipath_add_binding(binding: SessionBinding) -> Result<()> {
    let manager = multipath_manager();
    let mut config = manager.get_config();
    config.bindings.push(binding);
    apply_multipath_config(config)?;
    log::info!("已添加会话绑定规则");
    Ok(())
}

/// 删除会话绑定规则
pub fn multipath_remove_binding(domain_pattern: &str) -> Result<()> {
    let manager = multipath_manager();
    let mut config = manager.get_config();
    config.bindings.retain(|b| b.domain_pattern != domain_pattern);
    apply_multipath_config(config)?;
    log::info!("已删除会话绑定规则: {}", domain_pattern);
    Ok(())
}

/// 获取预定义的会话绑定规则
pub fn multipath_get_predefined_bindings() -> Vec<SessionBinding> {
    SessionBinding::all_predefined()
}

/// 添加节点池
pub fn multipath_add_pool(pool: NodePool) -> Result<()> {
    let manager = multipath_manager();
    let mut config = manager.get_config();
    config.node_pools.push(pool);
    apply_multipath_config(config)?;
    log::info!("已添加节点池");
    Ok(())
}

/// 删除节点池
pub fn multipath_remove_pool(pool_name: &str) -> Result<()> {
    let manager = multipath_manager();
    let mut config = manager.get_config();
    config.node_pools.retain(|p| p.name != pool_name);
    apply_multipath_config(config)?;
    log::info!("已删除节点池: {}", pool_name);
    Ok(())
}

/// 更新节点池
pub fn multipath_update_pool(pool: NodePool) -> Result<()> {
    let manager = multipath_manager();
    let mut config = manager.get_config();

    if let Some(existing) = config.node_pools.iter_mut().find(|p| p.name == pool.name) {
        *existing = pool;
        apply_multipath_config(config)?;
        log::info!("已更新节点池");
        Ok(())
    } else {
        Err(anyhow::anyhow!("节点池不存在"))
    }
}

/// 添加节点到池
pub fn multipath_add_node(pool_name: &str, node: PathNode) -> Result<()> {
    let manager = multipath_manager();
    let mut config = manager.get_config();

    if let Some(pool) = config.node_pools.iter_mut().find(|p| p.name == pool_name) {
        pool.nodes.push(node);
        apply_multipath_config(config)?;
        log::info!("已添加节点到池: {}", pool_name);
        Ok(())
    } else {
        Err(anyhow::anyhow!("节点池不存在"))
    }
}

/// 从池中删除节点
pub fn multipath_remove_node(pool_name: &str, node_name: &str) -> Result<()> {
    let manager = multipath_manager();
    let mut config = manager.get_config();

    if let Some(pool) = config.node_pools.iter_mut().find(|p| p.name == pool_name) {
        pool.nodes.retain(|n| n.name != node_name);
        apply_multipath_config(config)?;
        log::info!("已从池 {} 删除节点: {}", pool_name, node_name);
        Ok(())
    } else {
        Err(anyhow::anyhow!("节点池不存在"))
    }
}

/// 批量导入节点
pub fn multipath_import_nodes(pool_name: &str, nodes_yaml: &str) -> Result<usize> {
    let manager = multipath_manager();
    let mut config = manager.get_config();

    let nodes: Vec<PathNode> = serde_yaml_ng::from_str(nodes_yaml)?;
    if let Some(pool) = config.node_pools.iter_mut().find(|p| p.name == pool_name) {
        let count = nodes.len();
        pool.nodes.extend(nodes);
        apply_multipath_config(config)?;
        log::info!("已导入 {} 个节点到池: {}", count, pool_name);
        Ok(count)
    } else {
        Err(anyhow::anyhow!("节点池不存在"))
    }
}

/// 导出节点配置
pub fn multipath_export_nodes(pool_name: &str) -> Result<std::string::String> {
    let manager = multipath_manager();
    let config = manager.get_config();

    if let Some(pool) = config.node_pools.iter().find(|p| p.name == pool_name) {
        Ok(serde_yaml_ng::to_string(&pool.nodes)?)
    } else {
        Err(anyhow::anyhow!("节点池不存在"))
    }
}

/// 获取推荐配置
pub fn multipath_get_recommended_config() -> MultipathConfig {
    MultipathConfig {
        enabled: true,
        strategy: SlicingStrategy::Weighted,
        node_pools: vec![
            NodePool {
                name: "流媒体专用".to_string(),
                pool_type: PoolType::Streaming,
                nodes: Vec::new(),
                enabled: true,
            },
            NodePool {
                name: "游戏专用".to_string(),
                pool_type: PoolType::Gaming,
                nodes: Vec::new(),
                enabled: true,
            },
            NodePool {
                name: "下载专用".to_string(),
                pool_type: PoolType::Download,
                nodes: Vec::new(),
                enabled: true,
            },
            NodePool {
                name: "通用池".to_string(),
                pool_type: PoolType::General,
                nodes: Vec::new(),
                enabled: true,
            },
        ],
        min_fragment_size: 1024,
        max_fragment_size: 65536,
        reassembly_timeout: 5000,
        session_persistence: true,
        bindings: SessionBinding::all_predefined(),
    }
}

/// 获取节点统计
pub fn multipath_get_node_stats() -> HashMap<String, NodeStats> {
    let manager = multipath_manager();
    manager.get_node_stats()
}

/// Select a node for a multipath session.
pub fn multipath_select_node(domain: &str, session_id: u64) -> Option<String> {
    let manager = multipath_manager();
    manager.select_node(domain, session_id)
}

pub fn multipath_record_connection_end(node_name: &str, bytes: u64) {
    let manager = multipath_manager();
    manager.record_connection_end(node_name, bytes);
}

pub fn multipath_record_latency(node_name: &str, latency_us: u64) {
    let manager = multipath_manager();
    manager.record_latency(node_name, latency_us);
}

pub fn multipath_record_error(node_name: &str) {
    let manager = multipath_manager();
    manager.record_error(node_name);
}

pub fn multipath_get_active_session_count() -> usize {
    let manager = multipath_manager();
    manager.active_session_count()
}

/// 清理过期会话
pub fn multipath_cleanup_sessions() {
    let manager = multipath_manager();
    manager.cleanup_expired_sessions();
    log::info!("[Multipath] expired sessions cleaned");
}
